File: bindings/imgui_bundle/demos_python/demos_node_editor/demo_romeo_and_juliet.py
import os.path

from imgui_bundle import immapp, imgui, imgui_node_editor as ed, ImVec4


class Lover:
    name: str
    node_id: ed.NodeId
    pin_loves: ed.PinId
    pin_hates: ed.PinId
    pin_in: ed.PinId

    def __init__(self, name: str) -> None:
        self.name = name
        self.node_id = ed.NodeId.create()
        self.pin_loves = ed.PinId.create()
        self.pin_hates = ed.PinId.create()
        self.pin_in = ed.PinId.create()

    def draw(self) -> None:
        ed.begin_node(self.node_id)
        ed.begin_pin(self.pin_in, ed.PinKind.input)
        imgui.text(self.name)
        ed.end_pin()
        ed.begin_pin(self.pin_loves, ed.PinKind.output)
        imgui.text("Loves")
        ed.end_pin()
        ed.begin_pin(self.pin_hates, ed.PinKind.output)
        imgui.text("Hates")
        ed.end_pin()
        ed.end_node()


class Tie:
    lover: Lover
    loved: Lover
    kind: str
    id: ed.LinkId

    def __init__(self, lover: Lover, kind: str, loved: Lover) -> None:
        self.id = ed.LinkId.create()
        self.lover = lover
        self.loved = loved
        self.kind = kind

    def draw(self) -> None:
        red = ImVec4(1.0, 0.3, 0.2, 1.0)  # type: ignore
        green = ImVec4(0.3, 0.9, 0.0, 1.0)  # type: ignore
        if self.kind == "loves":
            ed.link(self.id, self.lover.pin_loves, self.loved.pin_in, green)
        else:
            ed.link(self.id, self.lover.pin_hates, self.loved.pin_in, red)


Romeo = Lover("Romeo")
Juliet = Lover("Juliet")
CountParis = Lover("Count Paris")
lovers = [Romeo, Juliet, CountParis]
links = [
    Tie(Romeo, "loves", Juliet),
    Tie(Juliet, "loves", Romeo),
    Tie(CountParis, "loves", Juliet),
    Tie(CountParis, "hates", Romeo),
    Tie(Romeo, "hates", CountParis),
]


def demo_gui():
    this_dir = os.path.dirname(__file__)
    ed.get_config().settings_file = this_dir + "/romeo_and_juliet.json"
    ed.begin("Romeo and Juliet")
    for lover in lovers:
        lover.draw()
    for link in links:
        link.draw()
    ed.end()


if __name__ == "__main__":
    immapp.run(demo_gui, with_node_editor=True, window_size=(1000, 800), window_title="It will not end well...")  # type: ignore

File: bindings/imgui_bundle/demos_cpp/demos_node_editor/demo_romeo_and_juliet.cpp
#ifdef IMGUI_BUNDLE_WITH_IMGUI_NODE_EDITOR
#include "imgui-node-editor/imgui_node_editor.h"
#include "imgui.h"

#include <string>
#include <memory>
#include <vector>


namespace ed = ax::NodeEditor;

namespace
{
    uintptr_t _next_id = 0;
    uintptr_t get_next_id() { return ++_next_id; }
}


struct Lover
{
    std::string name;
    ed::NodeId nodeId;
    ed::PinId pinLoves;
    ed::PinId pinHates;
    ed::PinId pinIn;

    explicit Lover(std::string _name) : name(std::move(_name))
    {
        nodeId = ed::NodeId(get_next_id());
        pinLoves = ed::PinId(get_next_id());
        pinHates = ed::PinId(get_next_id());
        pinIn = ed::PinId(get_next_id());
    }

    void Draw() const
    {
        ed::BeginNode(nodeId);
        ed::BeginPin(pinIn, ed::PinKind::Input);
        ImGui::Text("%s", name.c_str());
        ed::EndPin();
        ed::BeginPin(pinLoves, ed::PinKind::Output);
        ImGui::Text("Loves");
        ed::EndPin();
        ed::BeginPin(pinHates, ed::PinKind::Output);
        ImGui::Text("Hates");
        ed::EndPin();
        ed::EndNode();
    }
};


struct Tie
{
    std::shared_ptr<Lover> lover;
    std::shared_ptr<Lover> loved;
    std::string kind;
    ed::LinkId id;

    Tie(std::shared_ptr<Lover> _lover, const char* _kind, std::shared_ptr<Lover> _loved)
    {
        auto& self = *this;
        self.id = ed::LinkId(get_next_id());
        self.lover = std::move(_lover);
        self.loved = std::move(_loved);
        self.kind = _kind;
    }

    void Draw() const
    {
        auto& self = *this;
        ImVec4 red(1.0, 0.3, 0.2, 1.0);
        ImVec4 green(0.3, 0.9, 0.0, 1.0);
        if (kind == "loves")
            ed::Link(self.id, self.lover->pinLoves, self.loved->pinIn, green);
        else
            ed::Link(self.id, self.lover->pinHates, self.loved->pinIn, red);
    }
};


void demo_romeo_and_juliet()
{
    static std::vector<std::shared_ptr<Lover>> lovers;
    static std::vector<Tie> links;

    if (lovers.empty())
    {
        auto Romeo = std::make_shared<Lover>("Romeo");
        auto Juliet = std::make_shared<Lover>("Juliet");
        auto CountParis = std::make_shared<Lover>("Count Paris");
        lovers = { Romeo, Juliet, CountParis };

        links = {
            {Romeo, "loves", Juliet},
            {Juliet, "loves", Romeo},

            {CountParis, "loves", Juliet},

            {CountParis, "hates", Romeo},
            {Romeo, "hates", CountParis}
        };
    }

    // ed::GetConfig().SettingsFile = "romeo_and_juliet.json"; // GetConfig() is const!
    ed::Begin("Romeo and Juliet");
    for (auto &lover: lovers)
        lover->Draw();
    for (auto &link: links)
        link.Draw();
    ed::End();
}

#else // #ifdef IMGUI_BUNDLE_WITH_IMGUI_NODE_EDITOR
#include "imgui.h"
void demo_romeo_and_juliet() { ImGui::Text("This demo requires ImGui Node Editor\n"); }
#endif